pow/src/main/java/tech/pegasys/artemis/DepositRequestManager.java

```diff
@@ -0,0 +1,200 @@
+        .thenCompose(
+            groupedDepositEventResponses -> {
+              Map<String, SafeFuture<EthBlock.Block>> neededBlocksByHash =
+                  getMapOfEthBlockFutures(groupedDepositEventResponses);
+              return SafeFuture.allOf(neededBlocksByHash.values().toArray(SafeFuture[]::new))
+                  .thenApply(
+                      done ->
+                          constructDepositsFromBlockEvents(
+                              neededBlocksByHash, groupedDepositEventResponses));
+            })
+        .thenAccept(
+            (depositsFromBlockEventList) ->
+                depositsFromBlockEventList.stream()
+                    .sorted(Comparator.comparing(DepositsFromBlockEvent::getBlockNumber))
+                    .forEachOrdered(this::publishDeposits));
+  }
+
+  private List<DepositsFromBlockEvent> constructDepositsFromBlockEvents(
+      Map<String, SafeFuture<EthBlock.Block>> blockFutureByBlockHash,
+      List<List<DepositContract.DepositEventEventResponse>> groupedDepositEventResponses) {
+    return groupedDepositEventResponses.stream()
+        .map(
+            groupedDepositEventResponse -> {
+              String blockHash = groupedDepositEventResponse.get(0).log.getBlockHash();
+              EthBlock.Block block =
+                  checkNotNull(blockFutureByBlockHash.get(blockHash).getNow(null));
+              return new DepositsFromBlockEvent(
+                  UnsignedLong.valueOf(block.getNumber()),
+                  Bytes32.fromHexString(block.getHash()),
+                  UnsignedLong.valueOf(block.getTimestamp()),
+                  groupedDepositEventResponse.stream()
+                      .map(Deposit::new)
+                      .collect(Collectors.toList()));
+            })
+        .collect(Collectors.toList());
+  }
+
+  private Map<String, SafeFuture<EthBlock.Block>> getMapOfEthBlockFutures(
+      List<List<DepositContract.DepositEventEventResponse>> groupedDepositEventResponses) {
+    return groupedDepositEventResponses.stream()
+        .map(groupOfSameBlockDeposits -> groupOfSameBlockDeposits.get(0).log.getBlockHash())
+        .collect(Collectors.toUnmodifiableMap(k -> k, this::getEthBlockFuture));
+  }
+
+  // TODO: pass in web3j as an argument and share this code with Eth1DataManager to avoid
+  // duplication
+  private SafeFuture<EthBlock.Block> getEthBlockFuture(String blockHash) {
+    return SafeFuture.of(web3j.ethGetBlockByHash(blockHash, false).sendAsync())
+        .thenApply(EthBlock::getBlock);
+  }
+
+  private List<List<DepositContract.DepositEventEventResponse>>
+      groupDepositEventResponsesByBlockNumber(
+          List<DepositContract.DepositEventEventResponse> events) {
+    return new ArrayList<>(
+        events.stream()
+            .collect(
+                Collectors.groupingBy(
+                    event -> event.log.getBlockNumber(), TreeMap::new, Collectors.toList()))
+            .values());
+  }
```

Comment:
It would actually be easier to work with the data if this return a map of block hash to list of deposit responses (ie `Map<String, List<DepositContract.DepositEventEventResponse>>`.  Currently it's grouping by number - in the absence of forks that gives the same result as grouping by block hash (it's arguably more correct anyway, but we definitely don't have forks in the responses to a call to eth_getLogs).

Then in `getMapOfEthBlockFutures` you can just iterate the `keySet()` to get the block details for each hash and in `constructDepositsFromBlockEvents` use the `entrySet()` so you have both the block hash and list of deposits.  That way you can avoid the need to call `groupOfSameBlockDeposits.get(0).log.getBlockHash()` to get the relevant block hash.

Reply:
Done.